File: 2130-maximum-twin-sum-of-a-linked-list/2130-maximum-twin-sum-of-a-linked-list.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public int pairSum(ListNode head) {
        
        // Using 2 pointer approach
        
        // Find mid
        ListNode slow = head;
        ListNode fast = head;
        
        while(fast != null && fast.next != null)
        {
            slow = slow.next;
            fast = fast.next.next;
        }
        
        // Reverse
        ListNode curr = slow;
        ListNode next = null;
        ListNode prev = null;
        
        while(curr != null)
        {
            next = curr.next;
            curr.next = prev;
            prev = curr;
            curr = next;
        }
        
        ListNode start = head;
        ListNode end = prev;
        int ans = Integer.MIN_VALUE;
        
        while(end != null)
        {
            ans = Math.max(ans, start.val + end.val);
            start = start.next;
            end = end.next;
        }
        
        return ans;
    }
}

File: 2130-maximum-twin-sum-of-a-linked-list/2130-maximum-twin-sum-of-a-linked-list.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def pairSum(self, head: Optional[ListNode]) -> int:
        
        # Get middle of linked list
        slow , fast = head, head
        while fast != None and fast.next != None:
            slow = slow.next
            fast = fast.next.next
        
        
        # Reverse second part of linked list
        curr = slow
        next, prev = None, None
        while curr != None:
            next = curr.next
            curr.next = prev
            prev = curr
            curr = next
            
        # Get max sum of pairs
        maxVal = 0
        while(prev != None):
            maxVal = max(maxVal, head.val + prev.val)
            head = head.next
            prev = prev.next
            
        return maxVal